### PR TITLE
State and tests

### PR DESCRIPTION
**Pull request by brentschmaltz**
Replacement https://github.com/aspnet/Security/pull/233
Changed logic for user-state
Added additional testing

### CODE REVIEW

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 1]
```diff
@@ -291,36 +303,39 @@ protected override AuthenticationTicket AuthenticateCore()
...
+                    properties = SetUserStateOnMessage(message);
+                    if (properties == null)
+                    {
+                        Logger.LogError(Resources.OIDCH_0005_MessageStateIsInvalid);
+                        return null;
+                    }
                 }
 
-                // devs will need to hook AuthenticationFailedNotification to avoid having 'raw' runtime errors displayed to users.
+                // if any of the error fields are set, return null
```
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Fix this comment.

**Comment by Tratcher** [comment]
Close your old PR if you're done with it.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 2]
```diff
@@ -23,6 +23,9 @@ public class OpenIdConnectAuthenticationHandler : AuthenticationHandler<OpenIdCo
     {
         private const string NonceProperty = "N";
         private const string UriSchemeDelimiter = "://";
+        // should we consider adding a first class property on AuthenticationProperties similar to RedirectUri.
+        // if we don't, then we may want to think about where to make this value public.
+        private const string UserState = "userstate";
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: You could put it in OpenIdConnectAuthenticationDefaults.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 3]
```diff
@@ -69,21 +72,20 @@ protected override void ApplyResponseGrant()
...
+                    message.PostLogoutRedirectUri = properties.RedirectUri;
                 }
                 else if (!string.IsNullOrWhiteSpace(Options.PostLogoutRedirectUri))
                 {
-                    openIdConnectMessage.PostLogoutRedirectUri = Options.PostLogoutRedirectUri;
+                    message.PostLogoutRedirectUri = Options.PostLogoutRedirectUri;
                 }
 
-                var notification = new RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
+                var notification = new RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options, message);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Will the compiler let you omit the generics now that you're passing everything in as a direct argument?

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 4]
```diff
@@ -69,21 +72,20 @@ protected override void ApplyResponseGrant()
...
+                if (Options.Notifications.RedirectToIdentityProvider != null)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Inconsistent null checking. You check it here but not in ApplyResponseChallenge. Didn't we decide to just use NotNull for these on the Notifications class so we wouldn't have to check it everywhere?

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 5]
```diff
@@ -174,18 +176,36 @@ protected override void ApplyResponseChallenge()
...
                 message.Nonce = Options.ProtocolValidator.GenerateNonce();
+            }
+
+            var redirectToIdentityProviderNotification =
+                new RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options, message);
+
+            await Options.Notifications.RedirectToIdentityProvider(redirectToIdentityProviderNotification);
+            if (redirectToIdentityProviderNotification.HandledResponse)
+            {
+                Logger.LogInformation(Resources.OIDCH_0034_RedirectToIdentityProviderNotificationHandledResponse);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: LogVerbose?

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 6]
```diff
@@ -174,18 +176,36 @@ protected override void ApplyResponseChallenge()
...
+                return;
+            }
+            else if (redirectToIdentityProviderNotification.Skipped)
+            {
+                Logger.LogInformation(Resources.OIDCH_0035_RedirectToIdentityProviderNotificationSkipped);
+                return;
+            }
+
+            if (!string.IsNullOrWhiteSpace(message.Nonce))
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Use IsNullOrEmpty everywhere, not IsNullOrWhiteSpace. @Eilon

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 7]
```diff
@@ -586,34 +605,19 @@ private string ReadNonceCookie(string nonce)
             return null;
         }
 
-        private AuthenticationProperties GetPropertiesFromState(string state)
+        private AuthenticationProperties SetUserStateOnMessage(OpenIdConnectMessage message)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: DetaingleUserStateAndProperties? :-)

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationNotifications.cs [hunk 8]
```diff
@@ -44,7 +46,19 @@ public OpenIdConnectAuthenticationNotifications()
         /// <summary>
         /// Invoked to manipulate redirects to the identity provider for SignIn, SignOut, or Challenge.
         /// </summary>
-        public Func<RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>, Task> RedirectToIdentityProvider { get; set; }
+        public Func<RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>, Task> RedirectToIdentityProvider
+        {
+            get { return _redirectToIdentityProvider; }
+            set
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationNotifications.cs]: NotNull

File: src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs [hunk 9]
```diff
@@ -1,16 +1,43 @@
...
 
-        public TMessage ProtocolMessage { get; set; }
+        /// <summary>
+        /// Gets or sets the <see cref="{TMessage}"/>.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">if 'value' is null.</exception>
+        public TMessage ProtocolMessage
+        {
+            get { return _message; }
+            set
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs]: NotNull

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs [hunk 10]
```diff
@@ -102,214 +58,207 @@ public void LoggingLevel()
...
+            {
+                var formater = new AuthenticationPropertiesFormaterKeyValue();
+                var properties = new AuthenticationProperties();
+                var dataset = new TheoryData<Action<OpenIdConnectAuthenticationOptions>, OpenIdConnectMessage>();
+
+                var message = new OpenIdConnectMessage();
+                message.State = UrlEncoder.Default.UrlEncode(formater.Protect(properties));
+                message.Code = Guid.NewGuid().ToString();
+                message.Parameters.Add(ExpectedStateParameter, null);
+                dataset.Add(SetStateOptions, message);
```
Comment (Tratcher) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: Add comments describing each test scenario.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 5 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Yeah the rule we have with logging level Info and higher (e.g. also Warning and Error) is that there should be at most a *few* of them per request. If this is really just for a dev to diagnose some odd behavior, then Verbose or Debug are more appropriate.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 6 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Yeah I don't think a whitespace nonce would ever happen here, even in an error case. The check for whitespace is merely more expensive for no added value. The intention of the code presumably is about a *missing* nonce, not a nonce that is provided but is all whitespace.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 2 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: I think that would work, will make that change.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 3 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: I will try and see what happens.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 5 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Thanks, I will review https://github.com/aspnet/Logging/wiki/Guidelines
and make adjustments accordingly.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 6 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: I am fine with making that adjustment.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationNotifications.cs (hunk 8 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationNotifications.cs]: Will do.

File: src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs (hunk 9 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs]: Will do.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs (hunk 10 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: Sure.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 4 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: I removed some code for the null checks from the previous PR.
I would like to have a comprehensive POR before any other modifications. Then perform a consistency pass.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 7 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: I think that is a good idea.

**Comment by brentschmaltz** [comment]
@Tratcher @Eilon can you guys have a look at the latest.
I think we should make the change for leaving the notifications null and only engaging if !null.

**Comment by PinpointTownes** [comment]
@brentschmaltz I'm not sure it's really worth it. Anyway, we'll probably switch to good old methods in a future PR to match the pattern used by the OAuth2 middleware, so you won't have to care about null values in the handler itself: https://github.com/aspnet/Security/pull/257#issuecomment-115818763

File: src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs (hunk 9 above)
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs]:
```
public TMessage ProtocolMessage { get; [param: NotNull] set; }
```

I'm not sure we should have a setter though:
* `OpenIdConnectMessage` is a fully mutable type, so replacing it adds no real value.
* You need cumbersome `message = notification.ProtocolMessage` everywhere to make sure you flow the notification replaced by the user.

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 11]
```diff
@@ -586,34 +621,13 @@ private string ReadNonceCookie(string nonce)
             return null;
         }
 
-        private AuthenticationProperties GetPropertiesFromState(string state)
+        private AuthenticationProperties SetUserStateOnMessage(OpenIdConnectMessage message, AuthenticationProperties properties)
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: This method seems rather trivial and is used only once in the handler. Maybe you can remove it?

Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: @PinpointTownes  can you help me understand about leaving the notification null.  I thinks there is a gain here, if the user hasn't set a notification, then no need to new up the Notification and call a no-op.

+1 on removing the mini-method: SetUserStateOnMessage

File: src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs (hunk 9 above)
Comment (brentschmaltz) [inline src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs]: @PinpointTownes Thanks for the syntax lesson :-)

If we don't have the setter, we can pass [NoNull] as a param.
How do you feel about that?

**Comment by brentschmaltz** [comment]
@Eilon this PR has been marked cla-required.  What do I need to do?

File: src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs (hunk 9 above)
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/Notifications/RedirectToIdentityProviderNotification.cs]: Having an `HttpContext` in the `ProtocolMessage` property is a weird thing. Maybe we should just allow null values in the constructor and remove the setter?

**Comment by Eilon** [comment]
@brentschmaltz I sent you mail about the CLA.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs [hunk 12]
```diff
@@ -102,214 +57,223 @@ public void LoggingLevel()
...
+            var errors = new List<Tuple<LogEntry, LogEntry>>();
+            var expectedLogs = LoggingUtilities.PopulateLogEntries(expectedLogIndexes);
+            var handler = new OpenIdConnectAuthenticationHandlerForTestingAuthenticate(EmptyTask, EmptyTask);
+            var loggerFactory = new ReturnsLoggerLoggerFactory(logLevel);
+            var server = CreateServer(new ConfigureOptions<OpenIdConnectAuthenticationOptions>(action), UrlEncoder.Default, loggerFactory, handler);
+
+            await server.CreateClient().PostAsync("http://localhost", new FormUrlEncodedContent(message.Parameters));
+            LoggingUtilities.CheckLogs(loggerFactory.Logger.Logs, expectedLogs, errors);
+            Debug.WriteLine(LoggingUtilities.LoggingErrors(errors));
+            Assert.True(errors.Count == 0, LoggingUtilities.LoggingErrors(errors));
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: Can use `Assert.Equal`.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs [hunk 13]
```diff
@@ -394,9 +356,42 @@ private static void SecurityTokenReceivedSkippedOptions(OpenIdConnectAuthenticat
                 };
         }
 
-        private static void SecurityTokenValidatedHandledOptions(OpenIdConnectAuthenticationOptions options)
+        private static void SecurityTokenValidatorCannotReadToken(OpenIdConnectAuthenticationOptions options)
+        {
+            AuthenticationErrorHandledOptions(options);
+            Mock<ISecurityTokenValidator> mockValidator = new Mock<ISecurityTokenValidator>();
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: Can `var` a lot of these.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs [hunk 14]
```diff
@@ -422,14 +417,31 @@ private static void SecurityTokenValidatedSkippedOptions(OpenIdConnectAuthentica
                 };
         }
 
-        #endregion
+        private static void StateNullOptions(OpenIdConnectAuthenticationOptions options)
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: All three of the methods here have the same implementation - is that deliberate?

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectMiddlewareTests.cs [hunk 15]
```diff
@@ -33,25 +29,31 @@ public class OpenIdConnectMiddlewareTests
     {
         static string noncePrefix = "OpenIdConnect." + "Nonce.";
         static string nonceDelimiter = ".";
+        const string Challenge = "/challenge";
+        const string ChallengeWithOutContext = "/challengeWithOutContext";
+        const string ChallengeWithProperties = "/challengeWithProperties";
+        const string DefaultHost = @"https://example.com";
+        const string DefaultAuthority = @"https://login.windows.net/common";
+        const string ExpectedAuthorizeRequest = @"https://expectedauthorize.com/signin";
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectMiddlewareTests.cs]: Only use `.example.com/org/net` for fake URLs.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+using Microsoft.Framework.Logging;
+using Microsoft.Framework.OptionsModel;
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Authentication.Tests.OpenIdConnect
+{
+    /// <summary>
+    /// This formatter creates an easy to read string of the format: "'key1' 'value1' ..."
+    /// </summary>
+    public class AuthenticationPropertiesFormaterKeyValue : ISecureDataFormat<AuthenticationProperties>
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs]: One top-level type per file.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+
+                _logEntries.Add(logEntry);
+                Debug.WriteLine(logEntry.ToString());
+            }
+        }
+
+        public List<LogEntry> Logs { get { return _logEntries; } }
+    }
+
+    public class ReturnsLoggerLoggerFactory : ILoggerFactory
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs]: Use a mock for these?

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs [hunk 18]
```diff
@@ -10,6 +16,8 @@ namespace Microsoft.AspNet.Authentication.Tests.OpenIdConnect
     /// </summary>
     public class TestUtilities
     {
+        public const string DefaultHost = @"http://localhost";
+
         public static bool AreEqual<T>(object obj1, object obj2, Func<object, object, bool> comparer = null) where T : class
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs]: How is this method different from the existing xUnit.net method? https://github.com/xunit/xunit/blob/master/src/xunit.assert/Asserts/EqualityAsserts.cs#L30

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs [hunk 19]
```diff
@@ -105,4 +108,194 @@ private static bool AreEqual(Exception exception1, Exception exception2)
             return AreEqual<Exception>(exception1.InnerException, exception2.InnerException);
         }
     }
+
+    /// <summary>
+    /// Used to set up different configurations of metadata for different tests
+    /// </summary>
+    public class ConfigurationManager
```
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs]: One top-level type per file

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs (hunk 13 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: fixed

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs (hunk 14 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: Yes, it helps in debugging. To see the action in the debugger.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectMiddlewareTests.cs (hunk 15 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectMiddlewareTests.cs]: updated.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs (hunk 16 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs]: updated

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs (hunk 17 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/ScenarioHelperTypes.cs]: I need an actual instance of InMemoryLogger created.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs (hunk 18 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs]: handles multiple types.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs (hunk 19 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs]: updated.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs (hunk 12 above)
Comment (brentschmaltz) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: yes but there is no overload that has a message.

**Comment by Tratcher** [comment]
Making progress. You'll need to rebase after all the changes that went into Security last week.

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs (hunk 14 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs]: But these methods all do the same thing... we don't generally write code that is just for the debugger. What does this do that a debugger watch doesn't do?

File: test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs (hunk 18 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/TestUtilities.cs]: The xUnit one is generic - it handles any two types.

**Comment by brentschmaltz** [comment]
@Tratcher there are too many diff's to rebase.I will have to craft a new PR.

**Comment by PinpointTownes** [comment]
@brentschmaltz why are you so afraid of rebasing? :smile: 
It's actually far easier to rebase (even when there are merge conflicts) than creating a whole new PR.

**Comment by brentschmaltz** [comment]
updated

**Pull request closed without merging**